src/transformers/models/focalnet/modeling_focalnet.py

```diff
@@ -0,0 +1,1013 @@
+                    use_layerscale=config.use_layerscale,
+                    layerscale_value=config.layerscale_value,
+                    use_post_layernorm=config.use_post_layernorm,
+                    use_post_layernorm_in_modulation=config.use_post_layernorm_in_modulation,
+                    normalize_modulator=config.normalize_modulator,
+                )
+                for _ in range(depth)
+            ]
+        )
+
+        if downsample is not None:
+            self.downsample = downsample(
+                image_size=input_resolution,
+                patch_size=2,
+                num_channels=dim,
+                embed_dim=out_dim,
+                add_norm=True,
+                use_conv_embed=config.use_conv_embed,
+                is_stem=False,
+            )
+        else:
+            self.downsample = None
+
+        self.pointing = False
+
+    def forward(self, hidden_states: torch.Tensor, input_dimensions: Tuple[int, int]) -> Tuple[torch.Tensor]:
+        height, width = input_dimensions
+        for layer_module in self.layers:
+            hidden_states = layer_module(hidden_states, input_dimensions)
+
+        hidden_states_before_downsampling = hidden_states
+        if self.downsample is not None:
+            H, W = input_dimensions
+            hidden_states = hidden_states.transpose(1, 2).reshape(hidden_states_before_downsampling.shape[0], -1, H, W)
+            hidden_states, output_dimensions = self.downsample(hidden_states)
+
+        else:
+            output_dimensions = (height, width, height, width)
+
+        stage_outputs = (hidden_states, hidden_states_before_downsampling, output_dimensions)
+
+        return stage_outputs
+
+
+class FocalNetEncoder(nn.Module):
+    def __init__(self, config, grid_size):
+        super().__init__()
+        self.num_stages = len(config.depths)
+        self.config = config
+
+        # stochastic depth
+        dpr = [
+            x.item() for x in torch.linspace(0, config.drop_path_rate, sum(config.depths))
+        ]  # stochastic depth decay rule
+
+        embed_dim = [config.embed_dim * (2**i) for i in range(self.num_stages)]
+
+        self.stages = nn.ModuleList(
+            [
+                FocalNetStage(
+                    config=config,
+                    dim=embed_dim[i_layer],
+                    out_dim=embed_dim[i_layer + 1] if (i_layer < self.num_stages - 1) else None,
+                    input_resolution=(grid_size[0] // (2**i_layer), grid_size[1] // (2**i_layer)),
+                    depth=config.depths[i_layer],
+                    drop=config.hidden_dropout_prob,
+                    drop_path=dpr[sum(config.depths[:i_layer]) : sum(config.depths[: i_layer + 1])],
+                    focal_level=config.focal_levels[i_layer],
+                    focal_window=config.focal_windows[i_layer],
+                    downsample=FocalNetPatchEmbeddings if (i_layer < self.num_stages - 1) else None,
```

Comment:
Reading this, pretty much everything can be taken from the config if we pass along the index of stage. In any case the `drop` parameter can be picked from the config.